core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -610,10 +614,63 @@ void releaseFetchLock() {
     }
 
     private void initialize() {
-        // Initialize the partition.
         log.debug("Initializing share partition: {}-{}", groupId, topicIdPartition);
+        // Initialize the share partition by reading the state from the persister.
+        ReadShareGroupStateResult response;
+        try {
+            response = persister.readState(new ReadShareGroupStateParameters.Builder()
+                .setGroupTopicPartitionData(new GroupTopicPartitionData.Builder<PartitionIdLeaderEpochData>()
+                    .setGroupId(this.groupId)
+                    .setTopicsData(Collections.singletonList(new TopicData<>(topicIdPartition.topicId(),
+                        Collections.singletonList(PartitionFactory.newPartitionIdLeaderEpochData(topicIdPartition.partition(), 0)))))
+                    .build())
+                .build()
+            ).get();
+        } catch (InterruptedException | ExecutionException e) {
+            log.error("Failed to initialize the share partition: {}-{}", groupId, topicIdPartition, e);
+            throw new IllegalStateException(String.format("Failed to initialize the share partition %s-%s", groupId, topicIdPartition), e);
+        }
+
+        if (response == null || response.topicsData() == null || response.topicsData().size() != 1) {
+            log.error("Failed to initialize the share partition: {}-{}. Invalid state found: {}.",
+                groupId, topicIdPartition, response);
+            throw new IllegalStateException(String.format("Failed to initialize the share partition %s-%s", groupId, topicIdPartition));
+        }
 
-        // TODO: Provide implementation to initialize the share partition.
+        TopicData<PartitionAllData> state = response.topicsData().get(0);
+        if (state.topicId() != topicIdPartition.topicId() || state.partitions().size() != 1
+            || state.partitions().get(0).partition() != topicIdPartition.partition()) {
+            log.error("Failed to initialize the share partition: {}-{}. Invalid topic partition response: {}.",
+                groupId, topicIdPartition, response);
+            throw new IllegalStateException(String.format("Failed to initialize the share partition %s-%s", groupId, topicIdPartition));
+        }
+
+        PartitionAllData partitionData = state.partitions().get(0);
+        // Set the state epoch and end offset from the persisted state.
+        startOffset = partitionData.startOffset() != -1 ? partitionData.startOffset() : 0;
```

Comment:
We need to check the error code first before using the result.

Reply:
Sorry, I didn't get that. The response of the API is checked above. Am I missing somethign here?

Reply:
We need to check `partitionData.errorCode()` before using `partitionData.stateBatches()`, right?

Reply:
Yeah, my bad. I will send a PR with test.

Reply:
Addressed here: https://github.com/apache/kafka/pull/16956